Content.Shared/GameObjects/Components/Mobs/SharedActionsComponent.cs

```diff
@@ -0,0 +1,911 @@
+        public ValueTuple<TimeSpan, TimeSpan>? Cooldown;
+        public bool IsAtInitialState => IsAtInitialStateExceptCooldown && !Cooldown.HasValue;
+        public bool IsAtInitialStateExceptCooldown => !Enabled && !ToggledOn;
+
+        /// <summary>
+        /// Creates an action state for the indicated type, defaulting to the
+        /// initial state.
+        /// </summary>
+        public ActionState(bool enabled = false, bool toggledOn = false, ValueTuple<TimeSpan, TimeSpan>? cooldown = null)
+        {
+            Enabled = enabled;
+            ToggledOn = toggledOn;
+            Cooldown = cooldown;
+        }
+
+        public bool IsOnCooldown(TimeSpan curTime)
+        {
+            if (Cooldown == null) return false;
+            return curTime < Cooldown.Value.Item2;
+        }
+        public bool IsOnCooldown(IGameTiming gameTiming)
+        {
+            return IsOnCooldown(gameTiming.CurTime);
+        }
+    }
+
+    [Serializable, NetSerializable]
+    public abstract class PerformActionMessage : ComponentMessage
+    {
+        public readonly ActionType ActionType;
+
+        public PerformActionMessage(ActionType actionType)
+        {
+            Directed = true;
+            ActionType = actionType;
+        }
+    }
+
+    [Serializable, NetSerializable]
+    public abstract class PerformItemActionMessage : ComponentMessage
+    {
+        public readonly ItemActionType ActionType;
+        public readonly EntityUid Item;
+
+        public PerformItemActionMessage(ItemActionType actionType, EntityUid item)
+        {
+            Directed = true;
+            ActionType = actionType;
+            Item = item;
+        }
+    }
+
+    /// <summary>
+    /// A message that tells server we want to run the instant action logic.
+    /// </summary>
+    [Serializable, NetSerializable]
+    public class PerformInstantActionMessage : PerformActionMessage
+    {
+        public PerformInstantActionMessage(ActionType actionType) : base(actionType)
+        {
+        }
+    }
+
+    /// <summary>
+    /// A message that tells server we want to run the instant action logic.
+    /// </summary>
+    [Serializable, NetSerializable]
+    public class PerformInstantItemActionMessage : PerformItemActionMessage
+    {
+        public PerformInstantItemActionMessage(ItemActionType actionType, EntityUid item) : base(actionType, item)
+        {
+        }
+    }
+
+    /// <summary>
+    /// A message that tells server we want to toggle the indicated action.
+    /// </summary>
+    [Serializable, NetSerializable]
+    public class PerformToggleActionMessage : PerformActionMessage
+    {
+        /// <summary>
+        /// True if we are trying to toggle the action on, false if trying to toggle it off.
+        /// </summary>
+        public readonly bool ToggleOn;
+
+        public PerformToggleActionMessage(ActionType actionType, bool toggleOn) : base(actionType)
+        {
+            ToggleOn = toggleOn;
+        }
```

Comment:
So what you can do for these to save bandwidth is have a message for on and a message for off. It makes the code clunkier but you save the extra bool being sent.